src/VisualStudio/Core/Def/Implementation/CodeLens/CodeLensCallbackListener.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.VisualStudio.Language.CodeLens;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServices.Remote;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Settings;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.CodeLens
+{
+    /// <summary>
+    /// This is used by new codelens API to get extra data from VS if it is needed.
+    /// </summary>
+    [Export(typeof(ICodeLensCallbackListener))]
+    internal class CodeLensCallbackListener : ICodeLensCallbackListener
+    {
+        private const int DefaultMaxSearchResultsValue = 99;
+
+        private const string CodeLensUserSettingsConfigPath = @"Text Editor\Global Options";
+        private const string CodeLensMaxSearchResults = nameof(CodeLensMaxSearchResults);
+
+        private readonly VisualStudioWorkspaceImpl _workspace;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly JoinableTaskContext _joinableTaskContext;
+
+        private int _maxSearchResults = int.MinValue;
+
+        [ImportingConstructor]
+        public CodeLensCallbackListener(
+            SVsServiceProvider serviceProvider,
+            JoinableTaskContext joinableTaskContext,
+            VisualStudioWorkspaceImpl workspace)
+        {
+            _serviceProvider = serviceProvider;
+            _joinableTaskContext = joinableTaskContext;
+            _workspace = workspace;
+        }
+
+        public async System.Threading.Tasks.Task SynchronizePrimaryWorkspaceAsync(CancellationToken cancellationToken)
+        {
+            var client = await _workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                throw new InvalidOperationException("remote host doesn't exist");
+            }
+
+            await _workspace.SynchronizePrimaryWorkspaceAsync(_workspace.CurrentSolution, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<string> GetHostGroupIdAsync(CancellationToken cancellationToken)
+        {
+            // in VS host, RemoteHostClient is always ServiceHubRemoteHostClient
+            var client = (ServiceHubRemoteHostClient)await _workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                throw new InvalidOperationException("remote host doesn't exist");
+            }
+
+            return client.HostGroupId;
```

Comment:
this feels like a leeky abstraction... Why does thsi feature need to know about this Id?

Reply:
so code lens OOP can connect Roslyn OOP